File: javascript/twelve-days/twelve-days.js
const DAYS = ['first', 'second', 'third', 'fourth', 'fifth', 'sixth', 'seventh', 'eighth', 'ninth', 'tenth', 'eleventh', 'twelfth'];
const GIFTS = [
    'a Partridge in a Pear Tree',
    'two Turtle Doves',
    'three French Hens',
    'four Calling Birds',
    'five Gold Rings',
    'six Geese-a-Laying',
    'seven Swans-a-Swimming',
    'eight Maids-a-Milking',
    'nine Ladies Dancing',
    'ten Lords-a-Leaping',
    'eleven Pipers Piping',
    'twelve Drummers Drumming'
];

export function recite(from, to = from) {
    return [...DAYS.keys()].slice(from - 1, to).map((i) => reciteOne(i)).join('\n');
}

function reciteOne(n) {
    let verse = `On the ${DAYS[n]} day of Christmas my true love gave to me: `;
    if (n > 0) {
        verse = `${verse}${[...DAYS.keys()].slice(1, n + 1).reverse().map((i) => GIFTS[i]).join(', ')}, and `;
    }
    verse = `${verse}${GIFTS[0]}.\n`;
    return verse;
}

File: ruby/twelve-days/twelve_days.rb
module TwelveDays
  LYRICS_INFO = [
    {nth: "first", gift: "a Partridge in a Pear Tree"},
    {nth: "second", gift: "two Turtle Doves"},
    {nth: "third", gift: "three French Hens"},
    {nth: "fourth", gift: "four Calling Birds"},
    {nth: "fifth", gift: "five Gold Rings"},
    {nth: "sixth", gift: "six Geese-a-Laying"},
    {nth: "seventh", gift: "seven Swans-a-Swimming"},
    {nth: "eighth", gift: "eight Maids-a-Milking"},
    {nth: "ninth", gift: "nine Ladies Dancing"},
    {nth: "tenth", gift: "ten Lords-a-Leaping"},
    {nth: "eleventh", gift: "eleven Pipers Piping"},
    {nth: "twelfth", gift: "twelve Drummers Drumming"}
  ]

  def self.gift(day)
    case day
    when 1
      LYRICS_INFO[0][:gift]
    when 2
      LYRICS_INFO[1][:gift] + ", and " + gift(1)
    else
      LYRICS_INFO[day - 1][:gift] + ", " + gift(day - 1)
    end
  end

  def self.lyric(day)
    "On the #{LYRICS_INFO[day - 1][:nth]} day of Christmas my true love gave to me: #{gift(day)}.\n"
  end

  def self.lyrics(days)
    days.map { |day| lyric(day) }.join("\n")
  end

  def self.song
    lyrics(1..LYRICS_INFO.length)
  end
end
